dot/core/service.go

```diff
@@ -285,14 +296,23 @@ func (s *Service) Stop() error {
 	defer s.lock.Unlock()
 
 	// close channel to network service and BABE service
-	if !s.closed {
+	// thread safe way to check closed status
+	if atomic.LoadUint32(&s.closed) == uint32(1) {
 		if s.msgSend != nil {
 			close(s.msgSend)
 		}
 		if s.isBabeAuthority {
 			close(s.babeKill)
 		}
-		s.closed = true
+
+		defer func() {
+			// thread safe way to change closed status
+			canUnlock := atomic.CompareAndSwapUint32(&s.closed, 1, 0)
+			if !canUnlock {
+				panic("[core] Error when trying to change Service status from started to stopped.")
```

Comment:
return error instead of panic

Reply:
That wont be possible since its inside a defer func....

Reply:
ok, log it then instead?

Reply:
logging it will not be as easy to identify, I would recommend to panic. But, as you wish.